hail/src/main/scala/is/hail/expr/ir/MatrixWriter.scala

```diff
@@ -39,6 +61,255 @@ case class MatrixNativeWriter(
+                      WriteMetadata(MakeArray(GetField(writeEmpty, "filePath")),
+                        RVDSpecWriter(s"$path/globals/globals", RVDSpecMaker(emptySpec, RVDPartitioner.unkeyed(1)))),
+                      WriteMetadata(MakeArray(GetField(writeGlobals, "filePath")),
+                        RVDSpecWriter(s"$path/globals/rows", RVDSpecMaker(globalSpec, RVDPartitioner.unkeyed(1)))),
+                      WriteMetadata(MakeArray(I64(1)), globalTableWriter),
+                      WriteMetadata(MakeArray(GetField(colInfo, "filePath")),
+                        RVDSpecWriter(s"$path/cols/rows", RVDSpecMaker(colSpec, RVDPartitioner.unkeyed(1)))),
+                      WriteMetadata(MakeArray(GetField(colInfo, "partitionCounts")), colTableWriter),
+                      bindIR(ToArray(mapIR(ToStream(partInfo)) { fc => GetField(fc, "filePath") })) { files =>
+                        Begin(FastIndexedSeq(
+                          WriteMetadata(files, RVDSpecWriter(s"$path/rows/rows", RVDSpecMaker(rowSpec, lowered.partitioner))),
+                          WriteMetadata(files, RVDSpecWriter(s"$path/entries/rows", RVDSpecMaker(entrySpec, RVDPartitioner.unkeyed(lowered.numPartitions))))))
+                      },
+                      bindIR(ToArray(mapIR(ToStream(partInfo)) { fc => GetField(fc, "partitionCounts") })) { counts =>
+                        Begin(FastIndexedSeq(
+                            WriteMetadata(counts, rowTableWriter),
+                            WriteMetadata(counts, entriesTableWriter),
+                            WriteMetadata(makestruct("cols" -> GetField(colInfo, "partitionCounts"), "rows" -> counts), matrixWriter)))
+                      }))
+                  }
+                })))))
+    }
+  }
+}
+
+case class SplitPartitionNativeWriter(
+  spec1: AbstractTypedCodecSpec, partPrefix1: String,
+  spec2: AbstractTypedCodecSpec, partPrefix2: String,
+  index: Option[(String, PStruct)], localDir: Option[String]) extends PartitionWriter {
+  def stageLocally: Boolean = localDir.isDefined
+  def hasIndex: Boolean = index.isDefined
+  val filenameType = PCanonicalString(required = true)
+  def pContextType = PCanonicalString()
+  def pResultType: PCanonicalStruct =
+    PCanonicalStruct(required=true, "filePath" -> filenameType, "partitionCounts" -> PInt64(required=true))
+
+  def ctxType: Type = TString
+  def returnType: Type = pResultType.virtualType
+  def returnPType(ctxType: PType, streamType: PStream): PType = pResultType
+
+  if (stageLocally)
+    throw new LowererUnsupportedOperation("stageLocally option not yet implemented")
+  def ifIndexed[T >: Null](obj: => T): T = if (hasIndex) obj else null
+
+  def consumeStream(
+    context: EmitCode,
+    eltType: PStruct,
+    mb: EmitMethodBuilder[_],
+    region: Value[Region],
+    stream: SizedStream): EmitCode = {
+    val enc1 = spec1.buildEmitEncoderF(eltType, mb.ecb, typeInfo[Long]) //(Value[Region], Value[T], Value[OutputBuffer]) => Code[Unit]
```

Comment:
Ah hah! No, I forgot that that was a thing. Actually this was meant to be just a thing for me because I always forget the type signature, but that means it's probably good to explicitly have as the type annotation.